File: tools/detail_monitor.py
#!/usr/bin/env python3
"""闲鱼详情接口监听 — 被动捕获 itemDO.soldCnt 等字段"""
import frida
import sys
import json
from datetime import datetime

JS_CODE = open(__import__('pathlib').Path(__file__).parent / "detail_monitor.js", "r").read()

# 记录已打印的 itemId，去重用
seen = set()

def on_message(msg, data):
    if msg["type"] == "send":
        payload = msg.get("payload", "")
        if not isinstance(payload, dict):
            return
        t = payload.get("type", "")

        if t == "detail":
            item_id = payload.get("itemId", "?")
            if item_id in seen:
                print(f"[DETAIL] {item_id} (重复) soldCnt={payload.get('soldCnt')}")
                return
            seen.add(item_id)

            print()
            print("=" * 60)
            print(f"  ★ 抓获详情数据! source={payload.get('source','?')}")
            print(f"  itemId : {item_id}")
            print(f"  title  : {payload.get('title', '')}")
            print(f"  时间   : {datetime.now().strftime('%H:%M:%S')}")
            print("=" * 60)
            print()
            print(">>> ★ 销量/热度字段 ★ <<<")
            print(f"  soldCnt        = {payload.get('soldCnt')}")
            print(f"  soldPrice      = {payload.get('soldPrice')}")
            print(f"  wantCnt        = {payload.get('wantCnt')}")
            print(f"  collectCnt     = {payload.get('collectCnt')}")
            print(f"  browseCnt      = {payload.get('browseCnt')}")
            print(f"  commentCount   = {payload.get('commentCount')}")
            print()
            print(">>> 价格字段 <<<")
            print(f"  price          = {payload.get('price')}")
            print(f"  oldPrice       = {payload.get('oldPrice')}")
            print()
            print(">>> sellerDO <<<")
            print(f"  nick              = {payload.get('sellerNick')}")
            print(f"  hasSoldNumInteger = {payload.get('hasSoldNumInteger')}")
            print(f"  sellerItemCount   = {payload.get('sellerItemCount')}")
            print()
            print(f"  bodyLen: {payload.get('bodyLen','?')}")
            print(f"  itemDO 字段数: {payload.get('itemDOKeys','?')}")
            print("=" * 60)
            print()

        elif t == "http_req":
            print(f"[HTTP] {payload.get('url','')[:130]}")
        elif t == "mtop_detail":
            print(f"[MTOP] {payload.get('api','')} len={payload.get('len','')}")
        elif t == "status":
            print(f"[STATUS] {payload['msg']}")
        elif t == "okhttp_detail":
            print(f"[OKHTTP] url={payload.get('url','')[:100]}")
    elif msg["type"] == "error":
        print(f"[ERROR] {msg}")

def main():
    device = frida.get_usb_device()
    print(f"设备: {device.name}")

    # Attach 到已运行的闲鱼进程
    for proc in device.enumerate_processes():
        if proc.name == "闲鱼":
            pid = proc.pid
            break
    else:
        print("错误: 找不到闲鱼进程，请在模拟器上手动打开闲鱼")
        return
    print(f"Attach 到闲鱼进程 PID={pid}")
    session = device.attach(pid)

    script = session.create_script(JS_CODE)
    script.on("message", on_message)
    script.load()

    print("=" * 50)
    print("  闲鱼详情监听已就绪")
    print("  请在闲鱼 App 中打开任意商品详情页")
    print("  按 Ctrl+C 退出")
    print("=" * 50)
    print()

    import time
    try:
        print("等待 60 秒，期间任何详情页请求都会被捕获...", flush=True)
        for _ in range(120):
            time.sleep(0.5)
    except KeyboardInterrupt:
        pass
    finally:
        print("\n退出监听")
        session.detach()

if __name__ == "__main__":
    main()

File: tools/detail_monitor.js
// detail_monitor.js v5 — CLI 模式专用，console.log 输出

console.log("v5 脚本开始加载...");

setTimeout(function() {
    if (typeof Java === "undefined") {
        console.log("Java 未就绪，退出");
        return;
    }
    console.log("Java 可用，开始 hook...");

    Java.perform(function() {
        console.log("进入 Java.perform...");

        // === Hook MtopLauncher.send — 拦截所有 API 请求 ===
        try {
            var MtopLauncher = Java.use("com.taobao.android.remoteobject.easy.MtopLauncher");
            MtopLauncher.send.implementation = function(protocol, callback) {
                var api = "";
                try { api = protocol.getApiNameAndVersion() || ""; } catch(e) {}
                console.log("[SEND] " + api);
                this.send(protocol, callback);
            };
            console.log("Hook 1: MtopLauncher.send OK");
        } catch(e) {
            console.log("Hook 1 失败: " + e);
        }

        // === Hook RemoteMtopCallback.onMtopReturn — 拦截响应 ===
        try {
            var RMC = Java.use("com.taobao.android.remoteobject.easy.RemoteMtopCallback");
            var orig = RMC.onMtopReturn;
            RMC.onMtopReturn.implementation = function(ctx, map, ret) {
                var api = "";
                try { api = ret.getApi() || ""; } catch(e) {}
                var r = orig.call(this, ctx, map, ret);

                console.log("[RETURN] " + api);

                // 详情接口 / shade 接口 特殊处理
                if (api.indexOf("detail") >= 0 || api.indexOf("shade") >= 0) {
                    try {
                        var data = ret.getData();
                        if (data) {
                            var FastJSON = Java.use("com.alibaba.fastjson.JSON");
                            var jsonStr = FastJSON.toJSONString(data);
                            var obj = JSON.parse(jsonStr);
                            var itemDO = obj.itemDO || (obj.data && obj.data.itemDO) || {};
                            var sellerDO = obj.sellerDO || (obj.data && obj.data.sellerDO) || {};

                            console.log("");
                            console.log("========== 详情数据 ==========");
                            console.log("itemId: " + itemDO.itemId);
                            console.log("title:  " + itemDO.title);
                            console.log("--- 关键字段 ---");
                            console.log("soldCnt        = " + itemDO.soldCnt);
                            console.log("soldPrice      = " + itemDO.soldPrice);
                            console.log("wantCnt        = " + itemDO.wantCnt);
                            console.log("collectCnt     = " + itemDO.collectCnt);
                            console.log("browseCnt      = " + itemDO.browseCnt);
                            console.log("commentCount   = " + itemDO.commentCount);
                            console.log("price          = " + itemDO.price);
                            console.log("oldPrice       = " + itemDO.oldPrice);
                            console.log("quantity       = " + itemDO.quantity);
                            console.log("itemStatus     = " + itemDO.itemStatus);
                            console.log("--- sellerDO ---");
                            console.log("nick             = " + sellerDO.nick);
                            console.log("hasSoldNumInteger= " + sellerDO.hasSoldNumInteger);
                            console.log("itemCount        = " + sellerDO.itemCount);
                            console.log("");
                            console.log("--- itemDO 全部字段 (" + Object.keys(itemDO).length + ") ---");
                            for (var k in itemDO) {
                                var vs = String(itemDO[k]);
                                if (vs.length > 200) vs = vs.substring(0, 200) + "...";
                                console.log("  " + k + " = " + vs);
                            }
                            console.log("原始 JSON: " + jsonStr.length + " bytes");
                            console.log("==============================");
                            console.log("");
                        }
                    } catch(e2) {
                        console.log("[DETAIL ERR] " + e2);
                    }
                }
                return r;
            };
            console.log("Hook 2: onMtopReturn OK");
        } catch(e) {
            console.log("Hook 2 失败: " + e);
        }

        console.log("全部 Hook 安装完成！等待详情页请求...");
    });
}, 2000);
